util/cmdgen.py

```diff
@@ -0,0 +1,163 @@
+
+    Args:
+        filepath (Path): The file to check.
+        dry_run (bool): If set, commands will be printed, but not executed.
+        update (bool): If set, when content is out of date it will be updated.
+
+    Returns:
+        bool: If niether dry_run or update are set, will return True if an update is required.
+    '''
+    rel_path = filepath.relative_to(REPO_ROOT)
+    modified = False
+    needs_updating = False
+    content = filepath.read_text()
+    pos = 0
+    while True:
+        # search next start marker
+        match_start = START_MARKER_PATTERN.search(content, pos)
+        if match_start is None:
+            break  # no more replacements to do
+        match_start_linum = 1 + content[0:match_start.start(0)].count("\n")
+        cmd = match_start.group('cmd').strip()
+
+        # search end marker after the start marker
+        match_end = END_MARKER_PATTERN.search(content, pos)
+        if match_end is None:
+            sys.exit(
+                f"Error in {rel_path}: start marker on L{match_start_linum} "
+                "missing a matching end marker!"
+            )
+
+        if dry_run:
+            logger.info(f"{rel_path}:L{match_start_linum}: `{cmd}`")
+            # don't run
+            pos = match_end.end(0)
+            continue
+
+        # Run the found-command in a subshell
+        res = subprocess.run(cmd, shell=True, capture_output=True)
+        if res.stderr:
+            logger.info(
+                f"{rel_path}:L{match_start_linum}: `{cmd}` "
+                f"output the following error messages:\n{res.stderr.decode()}"
+            )
+        if res.returncode != 0:
+            sys.exit(
+                f"{rel_path}:L{match_start_linum}: `{cmd}` "
+                "had a non-zero return code of {res.returncode}."
+            )
+
+        pos = match_end.end(0)
+
+        new_content = '\n' + res.stdout.decode() + '\n'
+        old_content = content[match_start.end(0):match_end.start(0)]
+        if new_content != old_content:
+            if update:
+                logger.info(
+                    f"{rel_path}:L{match_start_linum}: Updating generated content."
+                )
+                # replace content
+                content = \
+                    content[0:match_start.end(0)] + \
+                    new_content + \
+                    content[match_end.start(0):]
+                modified = True
+                # update position to account for new content size
+                pos += len(new_content) - len(old_content)
+            else:
+                logger.info(
+                    f"{rel_path}:L{match_start_linum}: Generated content needs updating."
+                )
+                needs_updating = True
+
+    # write back
+    if modified:
+        filepath.write_text(content)
+
+    return needs_updating
+
+
+def main() -> int:
+    '''Run CMDGEN for all markdown files in the OpenTitan project.'''
```

Comment:
I don’t think the tool only applies to .md files with the globbing argument as-is? Is being more generic or more specific desirable here?
I think we should either 1) update the comment to not mention markdown specifically, or 2) add a check that excludes all non-markdown files from being updated. WDYT?

Reply:
I've just updated the comment for now.